src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenClosureLambdaTests.vb

```diff
@@ -3596,6 +3596,297 @@ End Class
                 }, c.GetMembers().Select(Function(member) member.ToString()))
             End Sub)
         End Sub
+
+        <Fact>
+        Public Sub DeclarationBlockClosures()
+            Dim source =
+<compilation>
+    <file name="a.vb">
+Imports System
+
+Class C
+    Sub New(a As Integer)
+        Dim f = Function() a
+    End Sub
+
+    Sub F1(a As Integer)
+        Dim f = Function() a
+    End Sub
+
+    Function F2(a As Integer) As Integer
+        Dim f = Function() a
+        Return 1
+    End Function
+
+    Property F3(a As Integer) As Integer
+        Get
+            Dim f = Function() a
+            Return 1
+        End Get
+
+        Set(value As Integer)
+            Dim f1 = Function() a
+            Dim f2 = Function() value
+        End Set
+    End Property
+
+    Custom Event F4 As Action
+        AddHandler(value As Action)
+            Dim f1 = Function() value
+        End AddHandler
+
+        RemoveHandler(value As Action)
+            Dim f1 = Function() value
+        End RemoveHandler
+
+        RaiseEvent()
+            Dim x = 1
+            Dim f1 = Function() x
+        End RaiseEvent
+    End Event
+
+    Shared Operator *(a As C, b As C) As C
+        Dim f1 = Function() a
+        Return a
+    End Operator
+
+    Shared Widening Operator CType(a As C) As Integer
+        Dim f1 = Function() a
+        Return 1
+    End Operator
+End Class    </file>
+</compilation>
+
+            CompileAndVerify(source)
```

Comment:
@gafter It compiles the code - I have put various assertions to lowering phase that would fail.